### PR TITLE
chore: refactor delete/update in merge into

### LINKED ISSUE #13454: bug: merge into the filter downcast value cause panic
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

v1.2.179-nightly

### What's Wrong?

merge into the filter downcast value cause panic
```
panicked at 'called `Option::unwrap()` on a `None` value', src/query/storages/fuse/src/operations/merge_into/processors/processor_merge_into_matched_and_split.rs:269:18
   0: common_tracing::panic_hook::log_panic
             at ./src/common/tracing/src/panic_hook.rs:35:21
   1: common_tracing::panic_hook::set_panic_hook::{{closure}}
             at ./src/common/tracing/src/panic_hook.rs:30:9
   2: call<(&core::panic::panic_info::PanicInfo), (dyn core::ops::function::Fn<(&core::panic::panic_info::PanicInfo), Output=()> + core::marker::Send + core::marker::Sync), alloc::alloc::Global>
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/alloc/src/boxed.rs:1999:9
   3: rust_panic_with_hook
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:709:13
   4: {closure#0}
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:595:13
   5: __rust_end_short_backtrace<std::panicking::begin_panic_handler::{closure_env#0}, !>
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/sys_common/backtrace.rs:151:18
   6: begin_panic_handler
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:593:5
   7: panic_fmt
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panicking.rs:67:14
   8: panic
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panicking.rs:117:5
   9: core::option::Option<T>::unwrap
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/option.rs:935:21
  10: <common_storages_fuse::operations::merge_into::processors::processor_merge_into_matched_and_split::MatchedSplitProcessor as common_pipeline_core::processors::processor::Processor>::process
             at ./src/query/storages/fuse/src/operations/merge_into/processors/processor_merge_into_matched_and_split.rs:265:46
  11: <alloc::boxed::Box<T> as common_pipeline_core::processors::processor::Processor>::process
             at ./src/query/pipeline/core/src/processors/processor.rs:169:9
  12: common_pipeline_core::processors::processor::ProcessorPtr::process
             at ./src/query/pipeline/core/src/processors/processor.rs:137:9
  13: databend_query::pipelines::executor::executor_worker_context::ExecutorWorkerContext::execute_sync_task
             at ./src/query/service/src/pipelines/executor/executor_worker_context.rs:82:9
  14: databend_query::pipelines::executor::executor_worker_context::ExecutorWorkerContext::execute_task
             at ./src/query/service/src/pipelines/executor/executor_worker_context.rs:73:46
  15: databend_query::pipelines::executor::pipeline_executor::PipelineExecutor::execute_single_thread
             at ./src/query/service/src/pipelines/executor/pipeline_executor.rs:381:45
  16: databend_query::pipelines::executor::pipeline_executor::PipelineExecutor::execute_threads::{{closure}}::{{closure}}
             at ./src/query/service/src/pipelines/executor/pipeline_executor.rs:337:27
  17: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panic/unwind_safe.rs:271:9
  18: std::panicking::try::do_call
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:500:40
  19: __rust_try
  20: std::panicking::try
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:464:19
  21: std::panic::catch_unwind
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panic.rs:142:14
  22: common_base::runtime::catch_unwind::catch_unwind
             at ./src/common/base/src/runtime/catch_unwind.rs:26:11
  23: databend_query::pipelines::executor::pipeline_executor::PipelineExecutor::execute_threads::{{closure}}
             at ./src/query/service/src/pipelines/executor/pipeline_executor.rs:336:34
  24: common_base::runtime::thread::Thread::named_spawn::{{closure}}
             at ./src/common/base/src/runtime/thread.rs:85:21
  25: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/sys_common/backtrace.rs:135:18
  26: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/thread/mod.rs:529:17
  27: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panic/unwind_safe.rs:271:9
  28: std::panicking::try::do_call
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:500:40
  29: __rust_try
  30: std::panicking::try
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:464:19
  31: std::thread::Builder::spawn_unchecked_::{{closure}}
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panic.rs:142:14
  32: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/ops/function.rs:250:5
  33: call_once<(), dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global>
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/alloc/src/boxed.rs:1985:9
  34: call_once<(), alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global>, alloc::alloc::Global>
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/alloc/src/boxed.rs:1985:9
  35: thread_start
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/sys/unix/thread.rs:108:17
  36: <unknown>
  37: <unknown>
```

### How to Reproduce?

```sql
mysql> create table tt1(a bool, b int);
Query OK, 0 rows affected (2.29 sec)

mysql> insert into tt1 values (true, 1), (false, 2);
Query OK, 2 rows affected (1.78 sec)

mysql> merge into tt1 using (select 1 as x) as tt2 on (2 > 1) when matched and a then delete;
ERROR 1105 (HY000): PanicError. Code: 1104, Text = called `Option::unwrap()` on a `None` value.
```

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by JackTan25**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR
1. fix #13454, #13298
2. refactor update/delete for merge into, remove 'into_column' usage, use Value<BooleanType> to avoid 
unwrap error.
3. add more multi_matched test
- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13631)

### CODE REVIEW

**Comment by JackTan25** [comment]
```sql
deploy@(distributed_metrics)/tanboyu_test> merge into target_table_cluster as t1 using (select * from source_table) as t2
on t1.l_partkey = t2.l_partkey and t1.l_orderkey = t2.l_orderkey and
t1.l_suppkey = t2.l_suppkey and t1.l_linenumber = t2.l_linenumber and t1.l_shipdate = t2.l_shipdate when
matched  then update * when not matched then insert *;

MERGE INTO target_table_cluster AS t1 USING (
  SELECT
    *
  FROM
    source_table
) AS t2 ON t1.l_partkey = t2.l_partkey
AND t1.l_orderkey = t2.l_orderkey
AND t1.l_suppkey = t2.l_suppkey
AND t1.l_linenumber = t2.l_linenumber
AND t1.l_shipdate = t2.l_shipdate
WHEN matched THEN
UPDATE
  *
  WHEN NOT matched THEN
INSERT
  *

0 row read in 24.399 sec. Processed 2.76 million row, 402.07 MiB (113.16 thousand row/s, 16.48 MiB/s)
```
performance down back. It seems not related to this pr.

File: src/query/storages/fuse/src/operations/merge_into/mutator/delete_by_expr_mutator.rs [hunk 1]
```diff
@@ -16,38 +16,94 @@ use common_exception::Result;
...
             expr,
             row_id_idx,
             func_ctx,
+            origin_input_columns,
         }
     }
 
-    // return block after delete, and the rowIds which are deleted
     pub fn delete_by_expr(&self, data_block: DataBlock) -> Result<(DataBlock, DataBlock)> {
+        // it's the first update, after update, we need to add a filter column
```
Comment (b41sh) [inline src/query/storages/fuse/src/operations/merge_into/mutator/delete_by_expr_mutator.rs:50]: This function is for delete, but the comment is update.

File: src/query/storages/fuse/src/operations/merge_into/mutator/delete_by_expr_mutator.rs [hunk 2]
```diff
@@ -16,38 +16,94 @@ use common_exception::Result;
...
+        if data_block.num_columns() == self.origin_input_columns {
+            // assert_eq!(expr.data_type(), &DataType::Boolean);
```
Comment (b41sh) [inline src/query/storages/fuse/src/operations/merge_into/mutator/delete_by_expr_mutator.rs:52]: Unused assert seems can be removed.

File: src/query/storages/fuse/src/operations/merge_into/mutator/delete_by_expr_mutator.rs [hunk 3]
```diff
@@ -16,38 +16,94 @@ use common_exception::Result;
...
+            &self.func_ctx,
+            data_block.num_rows(),
+            &BUILTIN_FUNCTIONS,
+        )?;
+        let filter_not = filter_not.try_downcast().unwrap();
+        Ok((old_filter, filter_not))
+    }
+
+    // return block after delete, and the rowIds which are deleted
+    pub fn delete_block(
```
Comment (b41sh) [inline src/query/storages/fuse/src/operations/merge_into/mutator/delete_by_expr_mutator.rs:80]: This function has some duplicate codes, may be we can refactor to make it more  clear.

**Comment by JackTan25** [comment]
```sql
deploy@(distributed_metrics)/tanboyu_test> merge into target_table_cluster as t1 using (select * from source_table) as t2
on t1.l_partkey = t2.l_partkey and t1.l_orderkey = t2.l_orderkey and
t1.l_suppkey = t2.l_suppkey and t1.l_linenumber = t2.l_linenumber and t1.l_shipdate = t2.l_shipdate when
matched  then update * when not matched then insert *;

MERGE INTO target_table_cluster AS t1 USING (
  SELECT
    *
  FROM
    source_table
) AS t2 ON t1.l_partkey = t2.l_partkey
AND t1.l_orderkey = t2.l_orderkey
AND t1.l_suppkey = t2.l_suppkey
AND t1.l_linenumber = t2.l_linenumber
AND t1.l_shipdate = t2.l_shipdate
WHEN matched THEN
UPDATE
  *
  WHEN NOT matched THEN
INSERT
  *

0 row read in 23.443 sec. Processed 2.93 million row, 426.67 MiB (125.09 thousand row/s, 18.20 MiB/s)

deploy@(distributed_metrics)/tanboyu_test> merge into target_table_cluster as t1 using (select * from source_table) as t2
on t1.l_partkey = t2.l_partkey and t1.l_orderkey = t2.l_orderkey and
t1.l_suppkey = t2.l_suppkey and t1.l_linenumber = t2.l_linenumber and t1.l_shipdate = t2.l_shipdate when
matched  then update * when not matched then insert *;

MERGE INTO target_table_cluster AS t1 USING (
  SELECT
    *
  FROM
    source_table
) AS t2 ON t1.l_partkey = t2.l_partkey
AND t1.l_orderkey = t2.l_orderkey
AND t1.l_suppkey = t2.l_suppkey
AND t1.l_linenumber = t2.l_linenumber
AND t1.l_shipdate = t2.l_shipdate
WHEN matched THEN
UPDATE
  *
  WHEN NOT matched THEN
INSERT
  *

0 row read in 9.550 sec. Processed 1.5 million row, 182.15 MiB (157.06 thousand row/s, 19.07 MiB/s)
```
well, when I execute multi times, it will be better, maybe the recluster is root cause. cc @dantengsky @SkyFan2002

File: src/query/storages/fuse/src/operations/merge_into/mutator/delete_by_expr_mutator.rs (hunk 3 above)
Comment (JackTan25) [reply src/query/storages/fuse/src/operations/merge_into/mutator/delete_by_expr_mutator.rs:80]: ok

**Comment by JackTan25** [comment]
extract common codes to util.rs for delete and update, reduce duplicate codes.

**Review by b41sh — APPROVED** [review]

**Pull request merged**